### PR TITLE
Refactor: bufferer

### LINKED ISSUE #26: Packet Underflow
When using the lib to call this rust server a error is thrown.

image

### PR DESCRIPTION
**Pull request by cainthebest**
A rework of `bufferer` to fix #26 and add optimizations

### CODE REVIEW

File: src/bufferer.rs [hunk 1]
```diff
@@ -1,294 +1,587 @@
...
+        u64_val: 0x47F12D29C7F74E9F,
+        i8_val: -126,
+        i16_val: -24368,
+        i32_val: -1874965832,
+        i64_val: -2049853289529583747,
+        f32_val: -0.6893411,
+        f64_val: 0.24812124701362135,
+    };
+
+    fn get_test_data<T>(f: fn(&TestData) -> T) -> T { f(&TEST_DATA) }
```
Comment (CosminPerRam) [inline src/bufferer.rs:292]: Why do we have this function? To make swapping test data more easily?

File: src/bufferer.rs [hunk 2]
```diff
@@ -1,294 +1,587 @@
...
+// ($deserializer:expr, optional_utf8, $length:expr) => {
+// $deserializer.read_optional_utf8_string($length)
+// };
+// ($deserializer:expr, remaining_utf8) => {
+// $deserializer.read_remaining_utf8_string()
+// };
+// ($deserializer:expr, utf16, $length:expr) => {
+// $deserializer.read_utf16_string($length)
+// };
+// }
```
Comment (CosminPerRam) [inline src/bufferer.rs:259]: The problem with getting strings (at least with utf8) is that we might want to:
1. get the remaining data as a string
2. get a string from the data (the end is a null byte)
3. look for a string, if there is one, return it, if not, dont (no null byte)

and i guess we can say that this was the initial incentive into rethinking the 'bufferer'.
This is just a note for your further strings implementation.

**Comment by CosminPerRam** [comment]
This draft looks very promising, thanks for pushing an idea for the rework, I'm looking forward to following progression.

File: src/bufferer.rs (hunk 1 above)
Comment (cainthebest) [reply src/bufferer.rs:292]: The function `get_test_data` returns the result of invoking `f` with a reference to the constant object `TEST_DATA` of type `TestData`. `get_test_data` provides a way to pass a function that operates on `TestData` objects, and then applies that function to a pre-defined `TestData` object. The return value of `get_test_data` will be the value returned by the function `f` when passed `TEST_DATA`. The main purpose of the `get_test_data` function is to provide a standardized way of accessing test data by encapsulating the test data in a constant object, and passing a function to operate on that data, `get_test_data` ensures that the test data is kept in a consistent and immutable state.

Comment (cainthebest) [reply src/bufferer.rs:292]: Its not actually needed but its there for utility use.

**Pull request closed without merging**